odxtools/diaglayer.py

```diff
@@ -416,24 +384,79 @@ def decode_response(self, response: Union[bytes, bytearray],
                 f"None of the services {possible_services} could parse {response.hex()}.")
         return decoded_messages
 
+    #####
+    # <communication parameter handling>
+    #####
+    def _compute_available_commmunication_parameters(self) -> List[CommunicationParameterRef]:
+        """Compute the list of communication parameters that apply to
+        the diagnostic layer
+
+        Be aware that the inheritance scheme for communication
+        parameters is slightly different than for objects that are
+        subject to value inheritance:
+
+        - The ODXLINK ID id of communication parameters is used to
+          override inherited parameters instead of the short name.
+        - A parameter is only overridden if the specified protocol
+          matches.
+
+        Note that the specification leaves some room for
+        interpretation here: It says that if no protocol is specified,
+        the parameter shall apply to any protocol. But what happens if
+        the the same comparam is specified with and without a
+        protocol? Is this allowed at all? If yes, which of these
+        definitions gets priority? How does this interact with
+        inheritance? The approach taken here is to allow such cases
+        and to use the specific comparams if possible whilst the ones
+        without a specified protocol are taken as fallbacks...
+
+        """
+        com_params_dict: Dict[Tuple[str, Optional[str]], CommunicationParameterRef] = dict()
+
+        # Look in parent refs for inherited communication
+        # parameters. First fetch the communication parameters from
+        # low priority parents, then update with increasing priority.
+        for parent_ref in self._get_parent_refs_sorted_by_priority():
+            for cp in parent_ref.layer._compute_available_commmunication_parameters():
+                com_params_dict[(cp.id_ref.ref_id, cp.protocol_snref)] = cp
+
+        # finally, handle the locally defined communication parameters
+        for cp in self.diag_layer_raw.communication_parameters:
+            com_params_dict[(cp.id_ref.ref_id, cp.protocol_snref)] = cp
+
+        return list(com_params_dict.values())
+
+    @property
+    def communication_parameters(self) -> List[CommunicationParameterRef]:
+        """All communication parameters applicable to this DiagLayer
+
+        Note that, although communication parameters use inheritance,
+        it is *not* the "value inheritance" scheme used by e.g. DOPs,
+        tables, state charts, ...
+        """
+        return self._communication_parameters
+
     def get_communication_parameter(
         self,
-        name: str,
+        cp_id: str,
```

Comment:
Is this ID or shortname?

Reply:
I have seems PDX files where the short name is CP_UniqueRespIdTable, the Diag Layer is called ISO_15765_2 BUT the XML REF and ID are NOT `ISO_15765_2.CP_UniqueRespIdTable`, rather they are randomly generated UUID from the modeling tool.

Reply:
I suppose we are in a pickle, then: there are parameters which exhibit the same short names (`CP_UniqueRespIdTable`, `CP_TesterPresentTime`, ...) yet, depending on the COMPARAM-SUBSET they were defined in (`ISO_15765_2` (CAN) or `ISO_13400_2_DIS_2015` (DoIP)), they exhibit either a completely different definition (`CP_UniqueRespIdTable`) or at least different semantics (`CP_TesterPresentTime`). I don't really like the current solution of relying on the `protocol_name` to hopefully retrieve the correct one, but if there are no better ideas, I guess we have to rely on that again. Any ideas?

Reply:
BTW: I consider the fact that the comparams are not fully standardized to be -- err -- unfortunate...

Reply:
alright: I reverted the comparam retrieval mechanism back to using short names: f8119a7ae22a1f51c4004b6a98b3448a9807de1b

This has the unfortunate consequence that two COMPARAMs with the same short name cannot be specified for the same diag_layer object (e.g. `CP_UniqueRespIdTable` for both CAN and DoIP). That said, for inheritance the ODXLINK IDs are still used, because the ID generation scheme of the editor does not matter here and I believe that this is "more correct".